src/daft-plan/src/planner.rs

```diff
@@ -145,41 +166,59 @@ pub fn plan(logical_plan: &LogicalPlan) -> DaftResult<PhysicalPlan> {
-                }
+            let input_partition_spec = input_physical.partition_spec();
+            let input_num_partitions = input_partition_spec.num_partitions;
+            let num_partitions = num_partitions.unwrap_or(input_num_partitions);
+            let repartitioned_plan = match scheme {
+                PartitionScheme::Unknown => match num_partitions.cmp(&input_num_partitions) {
+                    Ordering::Greater => {
+                        // Split input partitions into num_partitions.
+                        let split_op = PhysicalPlan::Split(Split::new(
+                            input_physical.clone(),
+                            input_num_partitions,
+                            num_partitions,
+                        ));
+                        PhysicalPlan::Flatten(Flatten::new(split_op.into()))
+                    }
+                    Ordering::Less => {
+                        // Coalesce input partitions into num_partitions.
+                        PhysicalPlan::Coalesce(Coalesce::new(
+                            input_physical.clone(),
+                            input_num_partitions,
+                            num_partitions,
+                        ))
+                    }
+                    Ordering::Equal => {
+                        // # of output partitions == # of input partitions, so we drop the redundant repartition.
+                        return Ok(input_physical.as_ref().clone());
+                    }
+                },
                 PartitionScheme::Random => {
                     let split_op = PhysicalPlan::FanoutRandom(FanoutRandom::new(
-                        input_physical,
-                        *num_partitions,
+                        input_physical.clone(),
+                        num_partitions,
                     ));
-                    Ok(PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into())))
+                    PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into()))
                 }
                 PartitionScheme::Hash => {
                     let split_op = PhysicalPlan::FanoutByHash(FanoutByHash::new(
-                        input_physical,
-                        *num_partitions,
+                        input_physical.clone(),
+                        num_partitions,
                         partition_by.clone(),
                     ));
-                    Ok(PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into())))
+                    PhysicalPlan::ReduceMerge(ReduceMerge::new(split_op.into()))
                 }
                 PartitionScheme::Range => unreachable!("Repartitioning by range is not supported"),
+            };
+            let repartitioned_partition_spec = repartitioned_plan.partition_spec();
+            if (input_num_partitions == 1 && repartitioned_partition_spec.num_partitions == 1)
+                || (repartitioned_partition_spec == input_partition_spec)
```

Comment:
@samster25 @jaychia As mentioned in Slack, the old optimizer had an additional guard here that `!matches!(repartition_partition_spec.schema, PartitionScheme::Range)`, i.e. that we wouldn't drop an apparently redundant repartition if it was using a range-based partitioning scheme. I'm not sure what the motivation was behind that, and we don't expose range-based repartitioning through the `df.repartition()` API or in any way that could produce a range-based partition spec for a logical `Repartition` operator, so for now I'm dropping that extra clause.

Lmk if y'all remember the motivation behind that!

Reply:
Resolved offline with a decision to remove it for now.